framework/src/utils/PerfGraph.C

```diff
@@ -0,0 +1,447 @@
+    case SELF_PERCENT:
+      return 100. * (section_it->second._self / app_time);
+    case CHILDREN_PERCENT:
+      return 100. * (section_it->second._children / app_time);
+    case TOTAL_PERCENT:
+      return 100. * (section_it->second._total / app_time);
+    default:
+      ::mooseError("Unknown TimeType");
+  }
+}
+
+void
+PerfGraph::push(const PerfID id)
+{
+  if (!_active)
+    return;
+
+  auto new_node = _stack[_current_position]->getChild(id);
+
+  // Set the start time
+  new_node->setStartTime(std::chrono::steady_clock::now());
+
+  // Increment the number of calls
+  new_node->incrementNumCalls();
+
+  _current_position++;
+
+  if (_current_position >= MAX_STACK_SIZE)
+    mooseError("PerfGraph is out of stack space!");
+
+  _stack[_current_position] = new_node;
+}
+
+void
+PerfGraph::pop()
+{
+  if (!_active)
+    return;
+
+  _stack[_current_position]->addTime(std::chrono::steady_clock::now());
+
+  _current_position--;
+}
+
+void
+PerfGraph::updateTiming()
+{
+  // First update all of the currently running nodes
+  auto now = std::chrono::steady_clock::now();
+  for (unsigned int i = 0; i <= _current_position; i++)
+  {
+    auto node = _stack[i];
+    node->addTime(now);
+    node->setStartTime(now);
+  }
+
+  // Zero out the entries
+  for (auto & section_time_it : _section_time)
+  {
+    auto & section_time = section_time_it.second;
+
+    section_time._num_calls = 0;
+    section_time._self = 0.;
+    section_time._children = 0.;
+    section_time._total = 0.;
+  }
+
+  recursivelyFillTime(_root_node.get());
+
+  // Update vector pointing to section times
+  // Note: we are doing this _after_ recursively filling
+  // because new entries may have been created
+  _section_time_ptrs.resize(_id_to_section_name.size());
+
+  for (auto & section_time_it : _section_time)
+  {
+    auto id = _section_name_to_id[section_time_it.first];
+
+    _section_time_ptrs[id] = &section_time_it.second;
+  }
+}
+
+void
+PerfGraph::recursivelyFillTime(PerfNode * current_node)
+{
+  auto id = current_node->id();
+
+  auto self = std::chrono::duration<double>(current_node->selfTime()).count();
+  auto children = std::chrono::duration<double>(current_node->childrenTime()).count();
+  auto total = std::chrono::duration<double>(current_node->totalTime()).count();
+  auto num_calls = current_node->numCalls();
+
+  // RHS insertion on purpose
+  auto & section_time = _section_time[_id_to_section_name[id]];
+
+  section_time._self += self;
+  section_time._children += children;
+  section_time._total += total;
+  section_time._num_calls += num_calls;
+
+  for (auto & child_it : current_node->children())
+    recursivelyFillTime(child_it.second.get());
+}
+
+void
+PerfGraph::recursivelyPrintGraph(PerfNode * current_node,
+                                 FullTable & vtable,
+                                 unsigned int level,
+                                 unsigned int current_depth)
+{
+  auto & name = _id_to_section_name[current_node->id()];
```

Comment:
`mooseAssert` or `find`, probably with the data structure below two. Yes, I realize that this is designed that they are in sync, but refactor does happen.

Reply:
I'll add an `assert`